015-kafka-topic-encryption.md

```diff
@@ -0,0 +1,169 @@
+If a topic matches a topic in the encryption policy, the contents of the _value_ field of each record is encrypted and replaced with the corresponding ciphertext. Metadata about the encryption (e.g., encrypted key or key reference, nonce) is added to the record's headers in order that this information is available later for decryption.
+
+The encryption metadata is currently stored in Kafka record headers. Record headers are optional name value/pairs stored in an array associated with the record. Record header names are not required to be unique, thus name "collisions" are possible. Ordering and versioning are used to avoid problems arising from names used for metadata which coincidentally equal names chosen by a Kafka client. Ordering means that the record header is rewritten such that the encryption-related headers
+occur first in the array. The first header always contains a version from which the exact
+metadata field names can be derived. The decryption algorithm therefore can discern between
+headers added by the encryption module and those added by the client. The version field
+is enables future migration and backwards compatibility should changes in semantics occur such as modified message formats or encryption options.
+
+Although not part of our reference implementation, the encryption of header values is potentially an additional requirement to consider. In this case, policy would be extended to indicate that specific or all client header values be encrypted. Since record headers are already processed during encryption, and since the algorithm can discriminate between encryption and client headers, it should be just an incremental extension encrypt header data as well. Encrypting header values will not impinge on storing encryption metadata in headers as only client headers will be encrypted. 
+Encrypting header names as well would likewise be a relatively simple extension however we recommend further investigating the full implications of encrypting header names.
+
+### Embedding Topic Encryption in a Proxy
+A proxy can be deployed as a free-standing process or as a sidecar in a Kubernetes pod. 
+In both cases, the proxy needs to intercept and de-serialize Kafka messages, encrypt/decrypt the relevant records, 
+and reserialize them into a new message. The overhead of a proxy potentially has an effect on performance.
+It also means that the proxy version must always be in phase with that of the broker.
+A proxy-based solution however has the advantage that both Kafka client and broker are unaware
+of the proxy and do not require any modification or configuration to support encryption at rest.
+
+Whether the proxy runs under the control of the client or the broker is considered a configuration
+issue, i.e. exactly the same proxy would be used in both cases, but when under the control of the
```

Comment:
I think there are some notable difference depending on where it is running which should be covered here:

* When running as a sidecar inside the broker pod, the proxy has to listen to a single port which will only get the connections for this particular broker. This can be also easily orchestrated and this port / address might be used as a advertised host in the broker configuration. And thus the proxy really needs to only deal with the _Produce_ and _Fetch_ requests.

* But on the client side, the advertised hosts metadata make it a way more complicated. They would essentially lead the clients to bypass the proxy and connect to the brokers directly to produce / fetch messages. So I think this would need some special handling, or? I guess there will be two options. One the proxy can just open more different ports (one per broker) and modify the advertised hosts in metadata. Or I guess it can approach it Istio-style and just high-jack all connections into the proxy. But that would need to be able to filter out non-Kafka connections and possibly also connections to other Kafka clusters which should not be encrypted.

Or did I missed some option?

Reply:
@scholzj No, you did not miss an option and yes this requires special handling. Kafka clients need a broker list and broker metadata which connects them through one or more proxies. Meanwhile, the brokers  need to interact with each other and answer metadata requests to their clients. This can be approached from different angles. We have viewed this as an integration issue and outside the scope of the core encryption proposal. Although we have deployed an implementation, I am unsure whether we are at liberty to provide details. I will clarify. In any case, this is an area where we very much welcome the expertise of the strimzi community. In the meantime, we will add a section to the proposal to call this out and possibly outline potential solutions.

Reply:
Sounds like a plan. I'm fine if there are not too many details ... but as you said, might be just worth mention either that it is out of scope or just outline the possible solutions and leave the details for implementation.